src/TodayEventMonitor.vala

```diff
@@ -0,0 +1,181 @@
+
+            return false;
+        });
+    }
+
+    private void on_events_updated (E.Source source, Gee.Collection<ECal.Component> events) {
+        Idle.add ( () => {
+            foreach (var event in events)
+                update_event (source, event);
+
+            return false;
+        });
+    }
+
+    private void on_events_removed (E.Source source, Gee.Collection<ECal.Component> events) {
+        Idle.add ( () => {
+            foreach (var event in events)
+                remove_event (source, event);
+
+            return false;
+        });
+    }
+
+    private void add_event (E.Source source, ECal.Component event) {
+        unowned ICal.Component comp = event.get_icalcomponent ();
+        debug ("Event [%s, %s, %s]".printf (comp.get_summary (), source.dup_display_name (), comp.get_uid ()));
+        foreach (var alarm_uid in event.get_alarm_uids ()) {
+            ECal.ComponentAlarm e_alarm = event.get_alarm (alarm_uid);
+            ECal.ComponentAlarmAction action;
+
+#if E_CAL_2_0
+            action = e_alarm.get_action ();
+#else
+            e_alarm.get_action (out action);
+#endif
+            switch (action) {
+                case (ECal.ComponentAlarmAction.DISPLAY):
+                    ECal.ComponentAlarmTrigger trigger;
+#if E_CAL_2_0
+                    trigger = e_alarm.get_trigger ();
+#else
+                    e_alarm.get_trigger (out trigger);
+#endif
+                    if (trigger.get_kind () == ECal.ComponentAlarmTriggerKind.RELATIVE_START) {
+                        ICal.Duration duration = trigger.get_duration ();
+                        var start_time = Calendar.Util.icaltime_to_datetime (comp.get_dtstart ());
+                        var now = new DateTime.now_local ();
+                        if (now.compare (start_time) > 0) {
+                            continue;
+                        }
+                        start_time = start_time.add_weeks (-(int)duration.get_weeks ()); //vala-lint=space-before-paren
+                        start_time = start_time.add_days (-(int)duration.get_days ()); //vala-lint=space-before-paren
+                        start_time = start_time.add_hours (-(int)duration.get_hours ()); //vala-lint=space-before-paren
+                        start_time = start_time.add_minutes (-(int)duration.get_minutes ()); //vala-lint=space-before-paren
+                        start_time = start_time.add_seconds (-(int)duration.get_seconds ()); //vala-lint=space-before-paren
+                        if (start_time.get_year () == now.get_year () && start_time.get_day_of_year () == now.get_day_of_year ()) {
+                            var time = time_until_now (start_time);
+                            if (time >= 0) {
+                                add_timeout.begin (source, event, (uint)time);
+                            }
+                        }
+                    }
+                    continue;
+                default:
+                    continue;
```

Comment:
Since the alarm has an action that we're not parsing, this would probably be a good case for a warning.

Reply:
Added a warning.